torch/ao/pruning/_experimental/pruner/README.md

```diff
@@ -1,93 +1,263 @@
+         saliency = -weights.norm(dim=tuple(range(1, weights.dim())), p=1)
+         num_to_pick = int(len(mask) * kwargs["sparsity_level"])
+         prune = saliency.topk(num_to_pick).indices
+
+         # Set the mask to be false for the rows we want to prune
+         mask.data[prune] = False
+
+# define patterns
+patterns = _get_default_structured_pruning_patterns()
+
+original = Model()
+# any configs passed in here are defaults that are propagated
+pruner = SaliencyPruner({"prune_bias": True}, patterns=patterns)
+pruner.prepare(original, sparse_config)
+pruner.enable_mask_update = True
+pruner.step()
+pruned_model = pruner.prune()
+```
 
-The Base Pruner inherits from the Base Sparsifier.
+Let's walk through this line by line to understand what the different parts are doing.
 
+###  Pruning Config
 
-# Motivation
+To specify the layers to prune we just need the fully qualified name (FQN) of the tensor you are looking to prune in the module.
+You can get the FQN of a tensor by printing out `model.named_parameters()`.
 
-Sparsifying weights allows us to skip some of the multiplications during the dot product (i.e. in the Linear layers), which ultimately translates into faster inference. With structured pruning, whole rows/columns of a tensor would be zeroed-out. This translates into model transformation (not just tensor transformation). Logically, the process of structured pruning is similar to removing some of the input/output channels in the layer completely.
+For example, for the model defined above, let's say that we want to prune the first Linear layer (Alpha)
 
-![prune logic](./images/prune_1.png)
+To do this we would define a config that looks like this:
 
+``` python
+pruning_config = [
+    {"tensor_fqn": "seq.0.weight"}
+]
+```
 
-# Design Choices
+To prune multiple layers, we just append entries to the pruning config.
+**tensor_fqn** is the only required key in the pruning config. You can pass additional information in the config, for example the sparsity level you want to prune to by adding a key to the config. You can then access this additional information when you update the masks.
 
+``` python
+pruning_config = [
+    {"tensor_fqn": "seq.0.weight", "sparsity_level": 0.7},
+    {"tensor_fqn": "seq.2.weight", "sparsity_level": 0.7},
+    {"tensor_fqn": "seq.4.weight", "sparsity_level": 0.7},
+    {"tensor_fqn": "linear.weight", "sparsity_level": 0.7},
+]
+```
 
-## Eager Mode
+### Implementing a Pruner
 
-**PruningParametrization:** After pruning, the shape of the weight changes (some of the output channels are pruned). That means the output of the current layer will have less output layers compared to the original. This means that the next layer should have less input channels.
+Now that we've defined the sparse config for our model, we need to specify the pruning criteria that we will use.
+To do this, we need to extend a `BaseStructuredSparsifier` with a custom `update_mask` function.
```

Comment:
I think ideally users will not need to implement their own Pruners, I'm not sure if this is necessary for an introduction